accepted/0000-selective-versions-resolutions.md

```diff
@@ -0,0 +1,423 @@
+
+Any potentially counter-intuitive situation will result in a warning being
+issued. This subject is discussed at the end of this section.
+
+## Examples
+
+We the following packages and their dependencies:
+```
+package-a@1.0.0 and 2.0.0
+ |_ package-d1@1.0.0
+     |_ package-d2@1.0.0
+
+package-b@1.0.0
+ |_ package-d1@2.0.0
+
+package-c@1.0.0
+ |_ package-a@2.0.0
+     |_ ...
+```
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-b": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-d1": "3.0.0"
+  }
+```
+
+yarn will use `package-d1@3.0.0` for every nested dependency to `package-d1`
+and will behave as expected with respect to the `node_modules` folder by not
+duplicating the `package-d1` installation.
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-b": "1.0.0"
+  },
+  "resolutions": {
+    "package-a/package-d1": "3.0.0"
+  }
+```
+
+yarn will use `package-d1@3.0.0` oly for `package-a` and `package-b` will
+still have `package-d1@2.0.0` in its own `node_modules`.
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-a": "3.0.0"
+  }
+```
+
+`package-a` will still be resolved to `1.0.0`, but `package-c` will have
+`package-a@3.0.0` in its own `node_modules`.
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "package-a": "3.0.0"
+  }
+```
+
+yarn will do nothing (see below why).
+
+With:
+```json
+ "devDependencies": {
+    "package-a": "1.0.0",
+    "package-c": "1.0.0"
+  },
+  "resolutions": {
+    "**/package-a/package-d1": "3.0.0"
+  }
+```
+
+yarn will use `package-d1@3.0.0` both for `package-a` and the nested
+dependency `package-a` of `package-c`.
+
+## Resolutions
+
+Each sub-field of the `resolutions` field is called a *resolution*.
+It is a JSON field expressed by two strings: the package designation on the
+left and a version specification on the right.
+
+### Package designation
+
+A *resolution* contains on the left-hand side a glob pattern applied to
+the dependency tree (and not to the `node_modules` directory tree, since the
+latter is the result of yarn resolution being influenced by the *resolution*).
+
+- `a/b` denotes the directly nested dependency `b` of the project's
+dependency `a`.
+- `**/a/b` denotes the directly nested dependency `b`
+of all the dependencies and nested dependencies `a` of the project.
+- `a` denotes the project's dependency `a` (see below for a discussion on
+the matter of specifying a *resolution* for one of the project dependency).
+- `a/**/b` denotes all the nested dependencies `b` of the project's
+dependency `a`.
+- `**/a` denotes all the nested dependencies `a` of the project.
+- `**/a-*` denotes all the nested dependencies of the project whose
+name starts with `a-`.
+- `**` denotes all the nested dependencies of the project (a bad idea mostly,

```

Comment:
I think we should limit usage of `**` and prohibit designations ending with `**` as I don't see the real use case for that scenario.